CODING-CONVENTIONS.md

```diff
@@ -0,0 +1,226 @@
+
+* Generally avoids interfaces with a single implementation unless they are explicitly being used to narrow the exposed API.
+
+* Uses the most general applicable type. For example,  `List` or `Collection` instead of `ArrayList`.
+
+## 3.3 Additional Design Elements
+
+Additional design elements include: 
+* Using Optional rather than returning null when not having a value is a normal case.
+* Considering exception and error handling as part of the overall design.   Pantheon avoids checked exceptions.
+* Giving threads are giving meaningful names. For example:
+ `Executors.newFixedThreadPool(1, new ThreadFactoryBuilder().setNameFormat(“Ibft”).build())`
+
+  
+# 4 Java 
+
+## 4.1 Style Guide 
+
+Pantheon follows the [Google code style](https://google.github.io/styleguide/javaguide.html) and uses spotless to ensure consistency of formatting. 
+
+To automatically reformat the code before creating a pull request, run: 
+
+```json
+./gradlew spotlessApply
+```
+
+### 4.1.1 Install Google Style Settings 
+
+To install the Google style settings in the Intellij IDE, download `intellij-java-google-style.xml` from [here](https://github.com/google/styleguide). 
+
+For Mac OS: 
+1. In Intellij, select **Preferences -> Editor -> Code Style**. 
+2. Select the small gear icon next to **Scheme** and select **Import Scheme → IntelliJ IDEA code style XML**.
+3. Select *GoogleStyle* as the coding style.
+
+For Windows: 
+1. Copy the downloaded file into the `config/codestyles` folder in your IntelliJ settings folder. 
+2. In Intellj, select **Settings -> Editor -> Code Style** and select the *google-styleguide* as the current code style. 
+
+## 4.2 Additional Java Style Guidelines
+
+## 4.2.1 Fields 
+
+Class-level fields are generally not separated by blank lines but can use a blank line to separate them into logical groupings.
+
+## 4.2.2 Final Keyword
+
+Method parameters must be final.  Class level and local fields should be final whenever possible.
+
+## 4.2.3 Common Methods
+
+* Getters follow idiomatic format with `get` prefix.  For example, `getBlock()` gets a block property.
+* Setters follow idiomatic format with `set` prefix. For example, `setBlock(Block block)` sets a block property.
+* For `toString methods`,  use the Guava 18+ `MoreObjects.toStringHelper`
+* Equals and `hashCode()` methods use the `Object.equals` and `Object.hash` methods (this is the _Java 7+_ template in IntelliJ. Don’t accept subclasses and don’t use getters)
+
+## 4.2.4 Testing
+
+* Don't use a fixed prefix (for example, `test`). Do explain the expected behaviour not just the situation. 
+  
+  Good: `returnTrueWhenValueIsXyz()`
+  Bad: `valueIsXyz()`
+
+* Use AssertJ for assertions in preference to JUnit’s Assert class.
+
+  To help future-proof the code (avoiding libraries that may be deprecated in the near future), avoid assertions from the `org.assertj.core.api.Java6Assertions` class. Java6 in the name is a concerning signal.
```

Comment:
I think this could be softend a bit.

'Prefer `org.assertj.core.api.Assertions` over the compatibility classes such as `org.assertj.core.api.Java6Assertions`'

Reply:
I'm not sure I've seen a case where `Java6Assertions` was actually useful - it's typically just accidentally selected or auto-imported incorrectly.